### PR TITLE
feat(v2): dark mode toggle customization

### PR DESCRIPTION
**Pull request by Drewbi**
## Motivation

Feature was requested in #3052 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Does it work

The approach I took allows users to include any css that would work in an inline react `style` prop. This disqualifies attributes such as `content` as it requires the ::before selector.
This approach allows users to customize the margin or other styling of any text or emoji they choose.

Toggle

## Docs

Added the new fields to the API `themeConfig` section of the `docusaurus.config.js` doc

```
themeConfig: {
    colorMode: {
      defaultMode: 'light',
      disableSwitch: false,
      respectPrefersColorScheme: true,
      switchConfig: {
        darkIcon: '🌙',
        darkIconStyle: {
          marginLeft: '2px',
        },
        lightIcon: '☀️',
        lightIconStyle: {
          marginLeft: '1px',
        },
      },
    },
```
I was considering including a link to [this](https://reactjs.org/docs/dom-elements.html#style) page where they explain about how react does inline css. Let me know if you think it needs it.

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/themeConfigSchema.js [hunk 1]
```diff
@@ -102,6 +102,15 @@ const ColorModeSchema = Joi.object({
...
+    darkIconStyle: Joi.object(),
+    lightIcon: Joi.string().default('🌞'),
+    lightIconStyle: Joi.object(),
+  }).default({
+    darkIcon: '🌜',
+    lightIcon: '🌞',
+  }),
 }).default({
   defaultMode: 'light',
   disableSwitch: false,
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/themeConfigSchema.js:116]: The e2e tests are failing

> (undefined) TypeError: Cannot destructure property `darkIcon` of 'undefined' or 'null'.



This is because your schema does not work if colorMode config is simply undefined. You need to include a `switchConfig` in the default object as well.

(As we have seen, Joi is not so idiomatic so I'd appreciate if you add validation for those edge cases in the configValidation.test.ts)

**Comment by slorber** [comment]
if that helps here's a pr where I validate algolia theme config: https://github.com/facebook/docusaurus/pull/3133/files

File: packages/docusaurus-theme-classic/src/themeConfigSchema.js (hunk 1 above)
Comment (slorber) [reply packages/docusaurus-theme-classic/src/themeConfigSchema.js:116]: @Drewbi the expected change is here, not in the component.

```
.default({
  defaultMode: 'light',
  disableSwitch: false,
  respectPrefersColorScheme: false,
  switchConfig: {
    darkIcon: '🌜',
    lightIcon: '🌞',
  },
```

This is weird to have to declare twice these  defaults for switchconfig but this is how Joi works, as it considers undefined and {} differently. @anshulrgoyal do you think there's a better way with Joi? (maybe an option I don't know)

File: packages/docusaurus-theme-classic/src/theme/Toggle/index.tsx [hunk 2]
```diff
@@ -13,17 +13,38 @@ import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
...
+  const {
+    themeConfig: {
+      colorMode: {
+        switchConfig: {
+          darkIcon = '🌜',
+          darkIconStyle = {},
+          lightIcon = '🌞',
+          lightIconStyle = {},
+        } = {},
+      } = {},
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Toggle/index.tsx:39]: We have introduced config/options/theme normalization exactly to avoid such code ^^ we don't want components to handle themselves default values, we want to always provide values to components in the first place

Comment (Drewbi) [reply packages/docusaurus-theme-classic/src/theme/Toggle/index.tsx:39]: Ah yep, did you want me to make a central DEFAULT_CONFIG for docusaurus-theme-classic in the same way you validated algolia's theme config?

Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/Toggle/index.tsx:39]: Not necessarily central, but you can make a `const DEFAULT_COLOR_MODE`

**Comment by Drewbi** [comment]
I opted to use `lodash/merge` as the nested config fields were being overwritten if only one of them had been changed. I hope the tests are adequate as well. I previously added the new fields to the API docs, should I add a section about the switch config to the guide as well?

File: packages/docusaurus-theme-classic/src/theme/Toggle/index.tsx [hunk 3]
```diff
@@ -13,17 +13,43 @@ import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
...
+      colorMode
+    },
+  } = siteConfig;
+  
+  const { switchConfig: {
+    darkIcon,
+    darkIconStyle,
+    lightIcon,
+    lightIconStyle
+  } } = mergeDefault(colorMode, DEFAULT_COLOR_MODE_CONFIG)
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Toggle/index.tsx:45]: the themeConfig should be provided to the components with default values already applied. We should not apply defaults in components

File: packages/docusaurus-theme-classic/src/validateThemeConfig.js [hunk 4]
```diff
@@ -6,6 +6,24 @@
...
+    darkIconStyle: {},
+    lightIcon: '🌞',
+    lightIconStyle: {},
+  },
+};
+exports.DEFAULT_COLOR_MODE_CONFIG = DEFAULT_COLOR_MODE_CONFIG;
+
+exports.mergeDefault = (config, defaultConfig) => {
+  return merge({}, defaultConfig, config);
+};
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/validateThemeConfig.js:26]: this method should be moved in tests, because we shouldn't need to merge defaults in production code, as Joi is supposed to do that job already

**Comment by slorber** [comment]
> I opted to use lodash/merge as the nested config fields were being overwritten if only one of them had been changed.

What do you mean? which nested config field exactly?

Can you add a test to cover the case you describe?

**Comment by Drewbi** [comment]
I've added some more comments to the documentation about the styling options.
I also think I finally understand how config validation works now. 
Hopefully should be good to go now.

**Comment by slorber** [comment]
thanks that looks good :)
